### PR TITLE
skiplist: doubly list [20x faster reverse scan]

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
Signed-off-by: zhangjinpeng1987 <zhangjinpeng@pingcap.com>

close https://github.com/tikv/agatedb/issues/134

Implement doubly link of skiplist for the single-thread write scenario. It makes the reverse scan 20x faster than before.
```
skiplist_reverse_scan   time:   [1.5644 ms]
doubly_skiplist_reverse_scan  time:   [75.162 us]
```

### CODE REVIEW

File: skiplist/src/list.rs [hunk 1]
```diff
@@ -431,8 +444,20 @@ impl<T: AsRef<Skiplist<C>>, C: KeyComparator> IterRef<T, C> {
 
     pub fn prev(&mut self) {
         assert!(self.valid());
-        unsafe {
-            self.cursor = self.list.as_ref().find_near(self.key(), true, false);
+        if self.list.as_ref().allow_concurrent_write {
+            unsafe {
+                self.cursor = self.list.as_ref().find_near(self.key(), true, false);
+            }
```
Comment (GanZiheng) [inline skiplist/src/list.rs:450]: Can we apply this improvement when concurrent write is allowed?

File: skiplist/benches/bench.rs [hunk 2]
```diff
@@ -163,10 +163,50 @@ fn bench_write_skiplist(c: &mut Criterion) {
...
+            let mut iter = list.iter_ref();
+            iter.seek_to_last();
+            while iter.valid() {
+                iter.prev();
+            }
+        })
+    });
+}
+
+fn bench_reverse_scan_skiplist_fast(c: &mut Criterion) {
```
Comment (Connor1996) [inline skiplist/benches/bench.rs:185]: what's the different with non-fast

File: skiplist/src/list.rs (hunk 1 above)
Comment (zhangjinpeng1987) [reply skiplist/src/list.rs:450]: It is much more complicated for concurrent write, ref to https://github.com/tikv/rocksdb/pull/278/files, maybe in the future we can port it.

**Review by Connor1996 — APPROVED** [review]
LGTM

File: skiplist/src/list.rs (hunk 1 above)
Comment (GanZiheng) [reply skiplist/src/list.rs:451]: I see, exactly not easy :joy:

**Pull request merged**